calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,860 @@
+        ir::Control::If(i) => {
+            count_barriers(&i.tbranch, map);
+            count_barriers(&i.fbranch, map);
+        }
+        _ => {}
+    }
+}
+
+fn fmt_shared_name(comp: &'static str, idx: &u64) -> ir::Id {
+    format!("{}_{}", comp, idx).into()
+}
+
+fn fmt_ind_name(
+    comp: &'static str,
+    idx_barrier: &u64,
+    idx_member: &u64,
+) -> ir::Id {
+    format!("{}_{}_{}", comp, idx_barrier, idx_member).into()
+}
+
+fn add_shared_primitive(
+    builder: &mut ir::Builder,
+    prefix: &'static str,
+    barrier_idx: &u64,
+    primitive: &str,
+    parameters: &[u64],
+    map: &mut HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Cell> {
+    let prim = builder.add_primitive(
+        fmt_shared_name(prefix, barrier_idx),
+        primitive,
+        parameters,
+    );
+    map.insert(
+        fmt_shared_name(prefix, barrier_idx),
+        prim.borrow().name().clone(),
+    );
+    prim
+}
+
+fn add_ind_primitive(
+    builder: &mut ir::Builder,
+    prefix: &'static str,
+    barrier_idx: &u64,
+    member_idx: &u64,
+    primitive: &str,
+    parameters: &[u64],
+    map: &mut HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Cell> {
+    let prim = builder.add_primitive(
+        fmt_ind_name(prefix, barrier_idx, member_idx),
+        primitive,
+        parameters,
+    );
+    map.insert(
+        fmt_ind_name(prefix, barrier_idx, member_idx),
+        prim.borrow().name().clone(),
+    );
+    prim
+}
+
+fn add_shared_group(
+    builder: &mut ir::Builder,
+    prefix: &'static str,
+    barrier_idx: &u64,
+    map: &mut HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Group> {
+    let group = builder.add_group(fmt_shared_name(prefix, barrier_idx));
+    map.insert(
+        fmt_shared_name(prefix, barrier_idx),
+        group.borrow().name().clone(),
+    );
+    group
+}
+
+fn add_ind_group(
+    builder: &mut ir::Builder,
+    prefix: &'static str,
+    barrier_idx: &u64,
+    member_idx: &u64,
+    map: &mut HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Group> {
+    let group =
+        builder.add_group(fmt_ind_name(prefix, barrier_idx, member_idx));
+    map.insert(
+        fmt_ind_name(prefix, barrier_idx, member_idx),
+        group.borrow().name().clone(),
+    );
+    group
+}
+
+fn add_assignment_to_group(
+    builder: &mut ir::Builder,
+    group: &RRC<ir::Group>,
+    dst: RRC<ir::Port>,
+    src: RRC<ir::Port>,
+    guard: ir::Guard,
+) {
+    let asmt = builder.build_assignment(dst, src, guard);
+    group.borrow_mut().assignments.push(asmt);
+}
+
+fn find_port(
+    comp: &mut ir::Component,
+    cell_name: &ir::Id,
+    port_name: &str,
+    map: &HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Port> {
+    comp.find_cell(map.get(cell_name).unwrap())
+        .unwrap()
+        .borrow()
+        .get(port_name)
+}
+
+fn find_group(
+    comp: &mut ir::Component,
+    group_name: &ir::Id,
+    map: &HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Group> {
+    comp.find_group(map.get(group_name).unwrap()).unwrap()
+}
+
+fn build_incr_barrier(
```

Comment:
Not a big fan of these methods. They seem bulky and hard to read. Let's chat synchronously about making them easier to write. I think using the macros will really help.